### ISSUE TITLE
TypeError: unorderable types: dict() < dict()

### ISSUE THREAD

**Issue by gcrieloue-main** [MycroftAI/adapt#19]
I got this error today : 

File "robot.py", line 73, in phase2
    result = process(text)
  File "robot.py", line 83, in process
    for intent in engine.determine_intent(text):
  File "/home/pi/Adapt/adapt/src/adapt-parser/adapt/engine.py", line 56, in determine_intent
    for result in parser.parse(utterance, N=num_results):
  File "/home/pi/Adapt/adapt/src/adapt-parser/adapt/parser.py", line 20, in parse
    tagged = self._tagger.tag(utterance.lower())
  File "/home/pi/Adapt/adapt/src/adapt-parser/adapt/entity_tagger.py", line 87, in tag
    entities = self._sort_and_merge_tags(entities)
  File "/home/pi/Adapt/adapt/src/adapt-parser/adapt/entity_tagger.py", line 32, in _sort_and_merge_tags
    decorated.sort()

**Comment by clusterfudge** [comment]
Hey hellsdark, can you provide some more details on your inputs to determine intent? It sounds like you may be passing in some dictionaries, which would most certainly explode.

**Comment by hellsdark** [comment]
Trying to have an example…

It seems to happen with this (not full, sorry) :

```
def __init__(self):
        self.intent = 'calculatorIntent'

        # define vocabulary
        self.calculator_keyword = [
                "combien",
                "calcule",
                "ajoute",
                "soustrait",
                "enlève"
                ]

        # structure intent
        self.calculator_intent = IntentBuilder(self.intent)\
                .require("CalculatorKeyword")\
                .require("Item")\
                .build()

    def register(self, engine):
        for wk in self.calculator_keyword:
            engine.register_entity(wk, "CalculatorKeyword")
        engine.register_intent_parser(self.calculator_intent)
        engine.register_regex_entity("Ajoute (?P<Item>\d+( ?[\+/\-\*] ?\d)*)")

```

and trying with engine.determine_intent("ajoute 20") .

The regex seems to work in a python interpreter.

**Comment by the-kid89** [comment]
Hello Hellsdark, your code is very incomplete which makes it really hard to determine what is causing the issue. Though from what you have provided it looks like you are building a class and you have no IntentDeterminationEngine, there is also nothing adding "Item" to your intent engine, and "CalculatorKeyword" is added but never created.

**Comment by hellsdark** [comment]
Well, I agree, I tried to simplify my issue because I have a lot of code (which overall works very well).
Actually, I have this problem only with some regex and some inputs, so I think it's on the regex there might be a problem (even if in simple python regex interpretation, it also works well…).

I have no time now to isolate the case. Maybe the bug should be close in the meantime.

**Comment by clusterfudge** [comment]
Hey @hellsdark , I'm glad to hear you're trying out adapt! But yeah, without something with which I can use to reproduce your error (in completion), I'm going to have a really hard time debugging with you. It may seem like a lot to task, but if you can create (or modify) a unit test in the existing tests to reproduce the error, it'd be much easier for me to step through that in a debugger and see what's going on.

**Comment by jdiemz** [comment]
Hi, I actually ran into this today, using version 0.2.7.

Same error.
```
(mycroft) [19:02:58][jdiemz@scarlet ~/projects/mycroft] $ ./test2.py snooze 1 for 20 minutes
Traceback (most recent call last):
  File "./test2.py", line 50, in <module>
    main()
  File "./test2.py", line 42, in main
    for intent in engine.determine_intent(' '.join(sys.argv[1:])):
  File "/home/jdiemz/projects/mycroft/src/adapt-parser/adapt/engine.py", line 57, in determine_intent
    for result in parser.parse(utterance, N=num_results):
  File "/home/jdiemz/projects/mycroft/src/adapt-parser/adapt/parser.py", line 20, in parse
    tagged = self._tagger.tag(utterance.lower())
  File "/home/jdiemz/projects/mycroft/src/adapt-parser/adapt/entity_tagger.py", line 87, in tag
    entities = self._sort_and_merge_tags(entities)
  File "/home/jdiemz/projects/mycroft/src/adapt-parser/adapt/entity_tagger.py", line 32, in _sort_and_merge_tags
    decorated.sort()
TypeError: unorderable types: dict() < dict()
```

Sample and output: https://gist.github.com/jdiemz/b39a3cbf882f4553c5467f5ba2463d31

I can resolve it by making the digit-matching regexes specific to the command... ie

```
engine.register_regex_entity('delete (?P<DeleteWhich>\d+)')
engine.register_regex_entity('snooze (?P<SnoozeWhich>\d+)')
```

So not sure if its an actual bug or not.

**Comment by clusterfudge** [comment]
Thanks for the gist, @jdiemz ! I was able to reproduce your error in python3.5 (but not 2.7, for some reason), and the fix was pretty straightforward. It's on the @MycroftAI team to merge and publish a new version, or you can take a source dependency on my fork for the time being (I have already merged into [master](https://github.com/clusterfudge/adapt)).

**Comment by aatchison** [comment]
Thanks @clusterfudge, I'll look at that tomorrow.

**Comment by forslund** [comment]
This seems to have been fixed by #44. Closing

**Comment by forslund** [comment]
This seems to have been fixed by #44. Closing
